### PR TITLE
Fix O(n^2) body buffering in `WSGIMiddleware` for large uploads

### LINKED ISSUE #22665: [BUG] Uploading 1GB model takes 5 minutes, with high CPU/memory usage
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### MLflow version

- Client: 3.11.1
- Tracking server: 3.11.1

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS 26.4.1
- **Python version**: 3.12

### Describe the problem

This is a reopen of https://github.com/mlflow/mlflow/issues/18696, verified against latest mlflow version.

When I upload 1GB file against mlflow server on localhost, it takes >5 minutes and mlflow memory consumption climbs to 30+ GB on my laptop (+heavy GC activity and heavy CPU usage)

While PR https://github.com/mlflow/mlflow/pull/19520 fixed the artifact download issue https://github.com/mlflow/mlflow/issues/19517 nicely, the model upload performance regression https://github.com/mlflow/mlflow/issues/18696 does not seem to be resolved.

The performance problem seems to be caused by the WSGIMiddleware, specifically this line:
https://github.com/Kludex/starlette/blob/edcc79eb6df0319c25aeeae77378a0220facf181/starlette/middleware/wsgi.py#L103

The bytes concatenation gets more and more expensive as the data size grows. In my debugging I observed data was received 256kB at a time. With a surgical patch to the WSGIMiddleware while loop to collect byte chunks first to a list and only join them after all chunks have been collected to contruct the full byte array, the performance improves dramatically and upload completes in 4 seconds.

However, appears the whole WSGIMiddleware is deprecated, so it may be that proposing patch to currently used WSGIMiddleware it is not the way to go:
https://github.com/Kludex/starlette/blob/edcc79eb6df0319c25aeeae77378a0220facf181/starlette/middleware/wsgi.py#L15 . Probably potential fix could be either to try the suggested alternative WSGIMiddleware, fork the WSGIMiddleware or complete the migration out of WSGIMiddleware assuming it is meant only as temporary approach in mlflow.

### Tracking information

```shell
System information: Darwin Darwin Kernel Version 25.4.0: Thu Mar 19 19:33:25 PDT 2026; root:xnu-12377.101.15~1/RELEASE_ARM64_T6041
Python version: 3.12.13
MLflow version: 3.11.1
Tracking URI: http://localhost:5000
Registry URI: http://localhost:5000
MLflow environment variables: 
  MLFLOW_TRACKING_URI: http://localhost:5000
MLflow dependencies: 
  Flask: 3.1.3
  Flask-CORS: 6.0.2
  Flask-WTF: 1.2.2
  aiohttp: 3.13.5
  alembic: 1.18.4
  boto3: 1.42.91
  botocore: 1.42.91
  click: 8.3.2
  cryptography: 46.0.7
  docker: 7.1.0
  fastapi: 0.136.0
  graphene: 3.4.3
  gunicorn: 25.3.0
  huey: 2.6.0
  kubernetes: 35.0.0
  matplotlib: 3.10.8
  mlflow-skinny: 3.11.1
  mlflow-tracing: 3.11.1
  numpy: 2.4.4
  pandas: 2.3.3
  psycopg2-binary: 2.9.11
  pyarrow: 23.0.1
  scikit-learn: 1.8.0
  scipy: 1.17.1
  skops: 0.13.0
  sqlalchemy: 2.0.49
  uvicorn: 0.44.0
```

### Code to reproduce issue

```python
import time

import mlflow
import mlflow.sklearn
import numpy as np
from sklearn.base import BaseEstimator, RegressorMixin
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler


class DummyLargeRegressor(BaseEstimator, RegressorMixin):
    def fit(self, X, y=None):
        self.random_payload_ = np.random.randint(0, 256, size=1000 * 1024 * 1024, dtype=np.uint8)
        return self

    def predict(self, X):
        return np.zeros(len(X))


def main():
    mlflow.set_tracking_uri("http://localhost:5000")

    X_train = np.random.randn(100, 10)
    y_train = np.random.randn(100)

    pipeline = Pipeline(
        [
            ("scaler", StandardScaler()),
            ("regressor", DummyLargeRegressor()),
        ]
    )

    pipeline.fit(X_train, y_train)

    print("Logging model to MLflow...")
    with mlflow.start_run() as run:
        start_time = time.time()
        mlflow.sklearn.log_model(pipeline, registered_model_name="registered")
        upload_time = time.time() - start_time
        run_id = run.info.run_id
        print(f"Model logged with run_id: {run_id}")
        print(f"Model upload took {upload_time:.2f} seconds")

    print("Downloading model from MLflow...")
    client = mlflow.MlflowClient()
    latest_version_info = client.search_model_versions(f"run_id='{run_id}'")[0]
    start_time = time.time()
    loaded_model = mlflow.sklearn.load_model(latest_version_info.source)
    download_time = time.time() - start_time
    print(f"Model download took {download_time:.2f} seconds")

    regressor = loaded_model.named_steps["regressor"]
    payload_size_mb = len(regressor.random_payload_) / (1024 * 1024)
    print(f"Downloaded model payload size: {payload_size_mb:.2f} MB")

if __name__ == "__main__":
    main()

```

### Stack trace

```
N/A
```

### Other info / logs

```
N/A
```

### What component(s) does this bug affect?

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

### PR DESCRIPTION
**Pull request by harupy**
### Related Issues/PRs
 
Closes #22665

### What changes are proposed in this pull request?

Starlette's `WSGIMiddleware` buffers the entire request body before passing it to Flask using `body += chunk` in a loop. For a 1GB upload arriving in ~4000 × 256KB chunks, this performs O(n^2) byte copies — explaining the 5+ minute upload times and 30GB memory spikes reported in #22665.

This PR replaces `WSGIMiddleware` with `_EfficientWSGIMiddleware` that collects chunks into a list and calls `b"".join()` once at the end — O(n) instead of O(n^2).

**Before (original `WSGIMiddleware`):** 1GB upload hangs/takes 5+ minutes
**After (`_EfficientWSGIMiddleware`):** 1GB upload completes in ~2 seconds

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Tested with a 1GB artifact upload against `mlflow server` on localhost:

```python
import os, tempfile, time, mlflow

mlflow.set_tracking_uri("http://127.0.0.1:5000")

with tempfile.NamedTemporaryFile(delete=False, suffix=".bin") as f:
    f.write(os.urandom(1000 * 1024 * 1024))
    tmp_path = f.name

with mlflow.start_run():
    start = time.time()
    mlflow.log_artifact(tmp_path)
    print(f"Upload took {time.time() - start:.2f}s")

os.unlink(tmp_path)
```

Result: `Upload took 2.08s (481.9 MB/s)`

### Does this PR require documentation update?

- [x] No.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] Yes. Fix O(n^2) body buffering in WSGIMiddleware that caused large artifact uploads (e.g., 1GB models) to take 5+ minutes with extreme memory usage. Uploads now complete in seconds.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

File: mlflow/server/fastapi_app.py [hunk 1]
```diff
@@ -34,6 +37,41 @@
...
+        chunks: list[bytes] = []
+        more_body = True
+        while more_body:
+            message = await receive()
+            chunk = message.get("body", b"")
+            if chunk:
+                chunks.append(chunk)
+            more_body = message.get("more_body", False)
+        body = b"".join(chunks)
+        environ = build_environ(self.scope, body)
```
Comment (hanlaur) [inline mlflow/server/fastapi_app.py:62]: I would recommend adding explicit `del chunks` right after joining the chunks to reduce number of copies in memory since `build_environ` will construct another BytesIO for the byte array.

Comment (harupy) [reply mlflow/server/fastapi_app.py:62]: @hanlaur Thanks for the comment. Applied your suggestion. Would you mind verifying this patch? I will too.

**Comment by harupy** [comment]
/review

---

✅ [Review](https://github.com/mlflow/mlflow/actions/runs/24783107443?pr=22729) completed
<details><summary>Review Output</summary>

PR #22729 approved. The fix correctly addresses the O(n^2) body buffering issue by using list accumulation + `b"".join()` — a minimal, well-scoped change with no regressions.

</details>

**Comment by kriscon-db** [comment]
Non-blocking thought on the deprecated-module coupling:

Is migrating to `a2wsgi` (what upstream points at as the replacement) on the near-term roadmap? If yes, this PR is a fine interim fix and the rest doesn't apply.

If not, how are we thinking about handling the starlette 0.47 release when it lands? Concretely: we're safe today only because `fastapi 0.115.14` requires `starlette>=0.40.0,<0.47.0` and `0.46.2` still ships the wsgi module. Once fastapi bumps its starlette cap past 0.47, transitive resolution will pull in a starlette that has removed the module and `mlflow server` will fail at import. Curious what the plan is for that window.

Side note: this PR also changes the import path from `fastapi.middleware.wsgi` to `starlette.middleware.wsgi` directly, so every `mlflow server` startup now emits `starlette.middleware.wsgi is deprecated and will be removed in a future release. Please refer to https://github.com/abersheeran/a2wsgi as a replacement.` on import. Flagging as a user-visible log regression independent of the import-time question above.

-- Claude-drafted, reviewed before posting.

**Comment by harupy** [comment]
@kriscon-db Good question. `a2wsgi` is solo-managed. We could use it but relying on it is also a risk. For long-term, migrating to FastAI is the best option, but it's a massive effort.

**Comment by kriscon-db** [comment]
> @kriscon-db Good question. `a2wsgi` is solo-managed. We could use it but relying on it is also a risk. For long-term, migrating to FastAI is the best option, but it's a massive effort.

Are we tracking that migration somewhere?  Mostly worried about what will happen in a few months around library bumps if we've punted the actual fix for this.

**Review by kriscon-db — APPROVED** [review]
LGTM as a short term fix.

**Comment by harupy** [comment]
No, we're not tracking that anywhere.

File: mlflow/server/fastapi_app.py (hunk 1 above)
Comment (hanlaur) [reply mlflow/server/fastapi_app.py:62]: Thanks @harupy. I did run the model upload script included in the PR and yes it works fast now after installing mlflow with pip from this PR branch.
